platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,181 @@
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
+
+        if (roundRosterPairs.size() > MAXIMUM_ROSTER_HISTORY_SIZE) {
+            final RoundRosterPair lastRemovedRoster = roundRosterPairs.removeLast();
+            removeRoster(lastRemovedRoster.activeRosterHash());
+
+            // At this phase of the implementation, the roster state has a fixed size limit for active rosters.
+            // Future implementations (e.g. DAB) can modify this.
+            if (roundRosterPairs.size() > MAXIMUM_ROSTER_HISTORY_SIZE) {
+                // additional safety check to ensure that the roster state does not contain more than set limit.
+                throw new IllegalStateException(
+                        "Active rosters in the Roster state cannot be more than  " + MAXIMUM_ROSTER_HISTORY_SIZE);
+            }
+        }
+
+        final Builder newRosterState = previousRosterState
+                .copyBuilder()
+                .candidateRosterHash(Bytes.EMPTY)
+                .roundRosterPairs(roundRosterPairs);
+        // since a new active roster is being set, the existing candidate roster is no longer valid
+        // so we remove it if it meets removal criteria.
+        removeRoster(previousRosterState.candidateRosterHash());
+        storeRoster(roster, activeRosterHash, newRosterState);
+    }
+
+    /**
+     * Returns the roster state or throws an exception if the state is null.
+     * @return the roster state
+     * @throws NullPointerException if the roster state is null
+     */
+    @NonNull
+    private RosterState rosterStateOrThrow() {
+        return requireNonNull(rosterState.get());
+    }
+
+    /**
+     * Stores the roster in the roster map and updates the roster store with the provided builder.
+     *
+     * @param rosterToStore              the roster to store
+     * @param rosterHash                   the hash of the roster
+     * @param rosterStateBuilder       the roster state builder
+     */
+    private void storeRoster(
+            @NonNull final Roster rosterToStore,
+            @NonNull final Bytes rosterHash,
+            @NonNull final Builder rosterStateBuilder) {
+
+        this.rosterState.put(rosterStateBuilder.build());
+        this.rosterMap.put(ProtoBytes.newBuilder().value(rosterHash).build(), rosterToStore);
+    }
+
+    /**
+     * Removes a roster from the roster map, but only if it doesn't match any of the active roster hashes in
+     * the roster state. The check ensures we don't inadvertently remove a roster still in use.
+     *
+     * @param rosterHash the hash of the roster
+     */
+    private void removeRoster(@NonNull final Bytes rosterHash) {
+        final List<RoundRosterPair> activeRosterHistory = rosterStateOrThrow().roundRosterPairs();
+        if (activeRosterHistory.stream()
+                .noneMatch(rosterPair -> rosterPair.activeRosterHash().equals(rosterHash))) {
```

Comment:
Should this be `anyMatch` ?

Reply:
I think the idea is we only want to remove a roster from the roster map if it's not used as an active roster in the round roster history.

Reply:
Ah yes makes sense. Thanks @derektriley